lib/web_ui/dev/build.dart

```diff
@@ -0,0 +1,188 @@
+      // Return a never-ending future.
+      return Completer<bool>().future;
+    } else {
+      return true;
+    }
+  }
+}
+
+Future<void> gn() {
+  print('Running gn...');
+  return runProcess(
+    path.join(environment.flutterDirectory.path, 'tools', 'gn'),
+    <String>[
+      '--unopt',
+      '--full-dart-sdk',
+    ],
+  );
+}
+
+// TODO(mdebbar): Make the ninja step interruptable in the pipeline.
+Future<void> ninja() {
+  print('Running ninja...');
+  return runProcess('ninja', <String>[
+    '-C',
+    environment.hostDebugUnoptDir.path,
+  ]);
+}
+
+enum PipelineStatus {
+  idle,
+  started,
+  stopping,
+  stopped,
+  error,
+  done,
+}
+
+typedef PipelineStep = Future<void> Function();
+
+class Pipeline {
+  Pipeline({@required this.steps});
+
+  final Iterable<PipelineStep> steps;
+
+  Future<dynamic> _currentStepFuture;
+
+  PipelineStatus status = PipelineStatus.idle;
+
+  Future<void> start() async {
+    status = PipelineStatus.started;
+    try {
+      for (PipelineStep step in steps) {
+        if (status != PipelineStatus.started) {
+          break;
+        }
+        _currentStepFuture = step();
+        await _currentStepFuture;
+      }
+      status = PipelineStatus.done;
+    } catch (_) {
+      status = PipelineStatus.error;
+    } finally {
+      _currentStepFuture = null;
+    }
+  }
+
+  Future<void> stop() {
+    status = PipelineStatus.stopping;
+    return (_currentStepFuture ?? Future<void>.value(null)).then((_) {
+      status = PipelineStatus.stopped;
+    });
+  }
+}
+
+class PipelineWatcher {
+  PipelineWatcher({
+    @required this.dir,
+    @required this.pipeline,
+  }) : watcher = DirectoryWatcher(dir);
+
+  /// The path of the directory to watch for changes.
+  final String dir;
+
+  /// The pipeline to be executed when an event is fired by the watcher.
+  final Pipeline pipeline;
+
+  /// Used to watch a directory for any file system changes.
+  final DirectoryWatcher watcher;
+
+  void start() {
+    watcher.events.listen(_onEvent);
+  }
+
+  int _pipelineRunCount = 0;
+  bool _pipelineScheduled = false;
+
+  void _onEvent(WatchEvent event) {
+    final String relativePath = path.relative(event.path, from: dir);
+    print('- [${event.type}] ${relativePath}');
+    if (_pipelineScheduled) {
+      return;
+    }
+    _pipelineRunCount++;
+    _pipelineScheduled = true;
+    Future.delayed(const Duration(milliseconds: 100)).then((_) {
+      _pipelineScheduled = false;
+      _runPipeline();
+    });
+  }
+
+  void _runPipeline() {
+    int runCount;
+    switch (pipeline.status) {
+      case PipelineStatus.started:
+        pipeline.stop().then((_) {
+          runCount = _pipelineRunCount;
+          return pipeline.start();
+        }).then((_) {
+          if (runCount == _pipelineRunCount) {
+            print('*** Done! ***');
+          }
+        });
+        break;
+
+      case PipelineStatus.stopping:
+        // We are already trying to stop the pipeline. No need to do anything.
```

Comment:
What will cause the pipeline to restart?

Reply:
The previous switch case does that.